### PR TITLE
Adding support for options when parsing

### PR DESCRIPTION
**Pull request by MrMeemus**
Addresses #384.

Options added:
- sql_dialect
  This option allows the user to select which sql dialect to use when parsing. Prior to this support, all SQL dialects (PGSQL, MYSQL, TSQL, etc.) would share the same list of regex to tokens mapping. This will lead to mismatch token types for certain words. 

     **Example:** TransactSQL statement **"@@Version"**
     Expected token type: Builtin **'@@version'**
     Actual token type: Operator **'@@'**, Identifier **'Version'**

     Currently supported values are **'TransactSQL'**. 
- additional_keywords
  This option allows the user to supply a list of additional strings to map to Keyword token type when parsing. Currently the list of keywords are hard-coded in the source code. This can create issues for users if they want to use the library against the latest version of their SQL dialect of choice that sqlparse may not have been updated with those new keywords.

### CODE REVIEW

File: sqlparse/keywords.py [hunk 1]
```diff
@@ -842,3 +915,513 @@ def is_keyword(value):
     'IN': tokens.Keyword,
     'LOOP': tokens.Keyword,
 }
+
+# Transact SQL Syntax.
+# See https://technet.microsoft.com/en-us/library/ms189822(v=sql.110).aspx
```
Comment (pensivebrian) [inline sqlparse/keywords.py]: SQL 2012? This list may be more up-to date:
[https://docs.microsoft.com/en-us/sql/t-sql/language-elements/reserved-keywords-transact-sql](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/reserved-keywords-transact-sql)

**Comment by MrMeemus** [comment]
@andialbrecht, what is the ETA for this PR? I'm working on updating mssql-cli which relies on this update to sqlparse. Please let me know if there is anything I can do to expedite the process. Thank you!

File: sqlparse/keywords.py [hunk 2]
```diff
@@ -77,15 +105,56 @@ def is_keyword(value):
...
+        (r'[<>=~!]+', tokens.Operator.Comparison),
+        (r'[+/@#%^&|?^-]+', tokens.Operator),
+    ]
+}
 
 FLAGS = re.IGNORECASE | re.UNICODE
-SQL_REGEX = [(re.compile(rx, FLAGS).match, tt) for rx, tt in SQL_REGEX['root']]
+
+# Leave empty so the client can add additional keywords to parse against via the parser options.
+ADDITIONAL_KEYWORDS = {}
```
Comment (andialbrecht) [inline sqlparse/keywords.py]: I think we should avoid a dictionary on module level that is changed by a function (update_additional_keywords). When using the module programmatically to parse multiple statements ADDITIONAL_KEYWORDS will not be reset after each run. So a first call with additional_keywords in options and a subsequent call or a different thread (both in the same process) without will still use the additional keywords from the first run.

Maybe we should treat additional keywords differently and not using them in the _is_keyword(value) functions. I'm not sure what side-effects a change of the _is_keyword() functions would introduce.

File: tests/test_grouping.py [hunk 3]
```diff
@@ -19,9 +21,11 @@ def test_grouping_parenthesis():
     assert len(parsed.tokens[2].tokens[3].tokens) == 3
 
 
-def test_grouping_comments():
+@pytest.mark.parametrize('options', [{'sql_dialect': 'Default'},
```
Comment (andialbrecht) [inline tests/test_grouping.py]: We should also add tests where options is an empty dictionary (the current behavior). OTOH 'Default' shouldn't be tested as it's just used internally when sql_dialect isn't given.

I'm not really happy how parametrize is used here. When we add other options than sql_dialect it gets really tricky to test all combination of options this way. But I'll have to peak into the pytest docs, there was a way to build such parameter matrices in pytest.

**Comment by MrMeemus** [comment]
@andialbrecht , thanks for the feedback! I will resume work on this PR in the coming weeks. I'm currently a bit back logged :)

**Comment by owst** [comment]
Hi @MrMeemus / @andialbrecht - is there any chance of this PR being progressed? I've been digging into https://github.com/dbcli/pgcli/issues/954 (no completion in `pgcli` given a table named `events`) and I've tracked the cause down to `sqlparse` [considering `events` as a keyword](https://github.com/andialbrecht/sqlparse/blob/28a2acdd9b307c95d2111f1f831b1a5afc634691/sqlparse/keywords.py#L749). However, `events` [is not a PostgreSQL keyword](https://www.postgresql.org/docs/current/sql-keywords-appendix.html) so it would be great if I could select a Postgres dialect to avoid the misclassification for `pgcli`'s purposes.

Thanks!
